libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/ChargingStationsMetadataUpdater.kt

```diff
@@ -0,0 +1,134 @@
+package com.mapbox.navigation.core.reroute
+
+import com.google.gson.JsonObject
+import com.google.gson.JsonPrimitive
+import com.mapbox.api.directions.v5.models.DirectionsResponse
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.base.route.RouterOrigin
+import com.mapbox.navigation.core.routeoptions.isEVRoute
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+// It's a temporary platform side workaround, the final solution will be on NN side.
+// See https://mapbox.atlassian.net/browse/NN-854
+internal suspend fun restoreChargingStationsMetadata(
+    originalRoute: NavigationRoute,
+    newRoutes: List<NavigationRoute>,
+    workerDispatcher: CoroutineDispatcher = Dispatchers.Default
+): List<NavigationRoute> = withContext(workerDispatcher) {
+    val newPrimaryRoute = newRoutes.first()
+    val rerouteRouteOptions = newPrimaryRoute.routeOptions
+    if (!rerouteRouteOptions.isEVRoute()) {
+        return@withContext newRoutes
+    }
+    val updatedResponseWithoutCorrectTypes = if (newPrimaryRoute.origin == RouterOrigin.Onboard) {
+        updateChargingStationsMetadataBasedOnRequestUrl(rerouteRouteOptions, newPrimaryRoute)
+    } else {
+        newPrimaryRoute.directionsResponse
+    }
+    val updatedResponse = updateChargingStationsTypes(
+        updatedResponseWithoutCorrectTypes,
+        originalRoute.getServerProvidedChargingStationsIds()
+    )
+    // TODO: optimise java memory usage creating a copy of a response NAVAND-1437
+    NavigationRoute.create(updatedResponse, rerouteRouteOptions, newPrimaryRoute.origin)
+}
+
+fun updateChargingStationsTypes(
+    response: DirectionsResponse,
+    serverProvidedStationsIds: Set<String>
+): DirectionsResponse {
+    val updatedRoutes = response.routes().map { route ->
+        val updatedWaypoints =
+            (route.waypoints() ?: emptyList()).mapIndexed { waypointIndex, waypoint ->
```

Comment:
Again, waypoints_per_route?

Reply:
EV routing requires `waypoints_per_route` to be `true`. Do you think it makes sense to support `false`?

Reply:
I think it is required only if alternatives are enabled.

Reply:
oh, it's required only if alternatives are enabled

Reply:
done